torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+
+      types.erase(new_end, types.end());
+    };
+
+    std::vector<TypePtr> lhs_types;
+    std::vector<TypePtr> rhs_types;
+
+    gather_types(obj, lhs_types, /*is_lhs=*/true);
+    gather_types(classinfo, rhs_types, /*is_lhs=*/false);
+
+    RefinementSet refinement;
+
+    TypePtr unified_true = nullptr;
+    TypePtr unified_false = nullptr;
+
+    if (obj.kind() == TK_VAR) {
+      std::string ident = Var(obj).name().name();
+
+      std::vector<Refinement> true_refinements;
+      std::vector<Refinement> false_refinements;
+
+      std::vector<TypePtr> isinstance_types;
+      std::vector<TypePtr> not_isinstance_types;
+
+      // We can discard any rhs types that we know statically would be
+      // impossible. For example, if we had:
+      //
+      //    def fn(x: Optional[str]):
+      //        if isinstance(x, (List[str], str, int)):
+      //            ...
+      //
+      // then `x` would be `str` in the true branch and `None` in the
+      // false branch, not `(List[str], str, int)` in the true branch
+      // and `None` in the false branch
+      for (TypePtr lhs_type : lhs_types) {
+        if (lhs_type == AnyType::get()) {
+          isinstance_types.insert(
+              isinstance_types.end(), rhs_types.begin(), rhs_types.end());
+          break;
         }
-        return false;
-      }
-      bool maybeOfKind(TypeKind kind, const TypePtr& actual_type) {
-        if (actual_type->kind() == AnyType::Kind) {
-          return true;
-        }
-        if (const auto optional_type = actual_type->cast<OptionalType>()) {
-          return optional_type->getElementType()->kind() == kind;
+        if (std::any_of(
+                rhs_types.begin(), rhs_types.end(), [&](TypePtr rhs_type) {
+                  return lhs_type->isSubtypeOf(rhs_type);
+                })) {
+          isinstance_types.push_back(lhs_type);
+        } else {
+          not_isinstance_types.push_back(lhs_type);
         }
-        if (const auto union_type = actual_type->cast<UnionType>()) {
+      }
+
+      // Get a single type for the true and false branches
+      unified_true = isinstance_types.empty() ? nullptr
+                                              : *unifyTypeList(
+                                                    isinstance_types,
+                                                    nowhere,
+                                                    /*default_to_union=*/true);
+      if (unified_true) {
+        true_refinements = {Refinement(ident, unified_true)};
+      }
+      unified_false = not_isinstance_types.empty()
+          ? nullptr
+          : *unifyTypeList(
+                not_isinstance_types,
+                nowhere,
+                /*default_to_union=*/true);
+      if (unified_false) {
+        false_refinements = {Refinement(ident, unified_false)};
+      }
+
+      refinement = RefinementSet(true_refinements, false_refinements);
+    }
+
+    bool is_statically_true =
```

Comment:
this seems like work that should have already been accomplished in our earlier pass. Do you mind explaining what didnt work  or was less preferable than current version about 
> for each type in the input, we want to see if it subtypes any of the types in isisntance:
for the types that do subtype types in isinstance, we record which types are their supertype, and add those supertypes to the runtime checks we need to generate in the isinstance. the union of all the types that are subtypes will be the type refinement in the true branch.
The union of types that do not subtype types in isinstance get added to the false branch type refinements
If either of the do/do not subtypes sets is exactly equal to the input then we know the isinstance check is statically true/false.

Reply:
> If either of the do/do not subtypes sets is exactly equal to the input then we know the isinstance check is statically true/false.

We're recreating work here that I think would already be done is all, but ya can leave for follow up if there is a cleanup that is possible.